sdk/include/reactos/rosctrls.h

```diff
@@ -581,3 +581,254 @@ class CTreeView :
     }
 
 };
+
+class CTooltips :
+    public CWindow
+{
+public: // Configuration methods
+
+    HWND Create(HWND hWndParent, DWORD dwStyles = 0, DWORD dwExStyles = 0)
+    {
+        if (!dwStyles)
+        {
+            dwStyles = WS_POPUP | TTS_NOPREFIX;
+        }
+
+        if (!dwExStyles)
+        {
+            dwExStyles = WS_EX_TOPMOST;
+        }
+
+        m_hWnd = CreateWindowExW(dwExStyles,
+            TOOLTIPS_CLASS,
+            TEXT(""),
+            dwStyles,
+            CW_USEDEFAULT, CW_USEDEFAULT,
+            CW_USEDEFAULT, CW_USEDEFAULT,
+            hWndParent,
+            NULL,
+            _AtlBaseModule.GetModuleInstance(),
+            NULL);
+
+        if (!m_hWnd)
+            return NULL;
+
+        SetWindowPos(HWND_TOPMOST, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOACTIVATE);
+
+        return m_hWnd;
+    }
+
+public: // Relay event
+
+        // Win7+: Can use GetMessageExtraInfo to provide the WPARAM value.
+    VOID RelayEvent(MSG * pMsg, WPARAM extraInfo = 0)
+    {
+        SendMessageW(TTM_RELAYEVENT, extraInfo, reinterpret_cast<LPARAM>(pMsg));
+    }
+
+public: // Helpers
+
+    INT GetToolCount()
+    {
+        return SendMessageW(TTM_GETTOOLCOUNT, 0, 0);
+    }
+
+    BOOL AddTool(IN CONST TTTOOLINFOW * pInfo)
+    {
+        return SendMessageW(TTM_ADDTOOL, 0, reinterpret_cast<LPARAM>(pInfo));
+    }
+
+    VOID DelTool(IN HWND hwndToolOwner, IN UINT uId)
+    {
+        TTTOOLINFOW info = { sizeof(TTTOOLINFOW), 0 };
+        info.hwnd = hwndToolOwner;
+        info.uId = uId;
+        SendMessageW(TTM_DELTOOL, 0, reinterpret_cast<LPARAM>(&info));
+    }
+
+    VOID NewToolRect(IN HWND hwndToolOwner, IN UINT uId, IN RECT rect)
+    {
+        TTTOOLINFOW info = { sizeof(TTTOOLINFOW), 0 };
+        info.hwnd = hwndToolOwner;
+        info.uId = uId;
+        info.rect = rect;
+        SendMessageW(TTM_NEWTOOLRECT, 0, reinterpret_cast<LPARAM>(&info));
+    }
+
+    BOOL GetToolInfo(IN HWND hwndToolOwner, IN UINT uId, OUT TTTOOLINFOW * pInfo)
+    {
+        pInfo->cbSize = sizeof(TTTOOLINFOW);
```

Comment:
ATLASSERT(pInfo);
Also, shouldn't the caller fill cbSize?

Reply:
I wanted to use "out" semantics, but you are right, in this case it doesn't make sense.

Reply:
then make it IN OUT